src/analysis/max_min_driving_force.jl

```diff
@@ -55,56 +55,95 @@ optimization problems.
 """
 function max_min_driving_force(
     model::MetabolicModel,
-    gibbs_free_energies::Dict{String,Float64},
+    metabolite_gibbs_free_energies::Dict{String,Float64},
     optimizer;
-    ignore_metabolites::Vector{String} = ["h", "h2o"],
+    flux_solution::Dict{String,Float64} = Dict{String,Float64}(),
+    reaction_gibbs_free_energy_adjustments::Dict{String,Float64} = Dict{String,Float64}(),
+    metabolite_gibbs_free_energy_errors::Dict{String,Float64} = Dict{String,Float64}(),
+    proton_ids::Vector{String} = ["h_c", "h_e"],
+    water_ids::Vector{String} = ["h2o_c", "h2o_e"],
     constant_concentrations::Dict{String,Float64} = Dict{String,Float64}(),
     concentration_ratios::Dict{Tuple{String,String},Float64} = Dict{
         Tuple{String,String},
         Float64,
     }(),
+    ignore_reactions::Vector{String} = Vector{String}(),
     concentration_lb = 1e-6,
     concentration_ub = 10e-3,
     T = _constants.T,
     R = _constants.R,
+    metabolite_gibbs_free_energy_error_fallback_fraction = 0.01,
     modifications = [],
+    negligible_flux = 1e-6,
 )
-    # get ΔGs in order of reactions
-    dg0s = [gibbs_free_energies[rid] for rid in reactions(model)]
 
-    # get all metabolites that participate in the ΔG calculations (no protons, no water)
-    midxs = [
-        midx for
-        (midx, mid) in enumerate(metabolites(model)) if !(mid in ignore_metabolites)
-    ] # idx in original model
-    mids = metabolites(model)[midxs]
-
-    # find the corresponding metabolites
-    S = stoichiometry(model)[midxs, :]
-
-    # start building the optimization model
     opt_model = Model(optimizer)
+
     @variables opt_model begin
         minDF
-        logcs[1:length(midxs)]
-        dgs[1:n_reactions(model)]
+        logcs[1:n_metabolites(model)]
+        dgfs[1:n_metabolites(model)]
+        dgrs[1:n_reactions(model)]
     end
 
-    @constraints opt_model begin
-        minDF .<= -dgs
-        dgs .<= 0
-        dgs .== dg0s .+ R * T .* S' * logcs
+    # get metabolite ΔG⁰ of formation in order of metabolites in the model
+    dgf0s = [metabolite_gibbs_free_energies[mid] for mid in metabolites(model)]
+
+    # set proton ΔG formation to zero (biochemical thermodynamic assumption)
+    proton_idxs = Int.(indexin(proton_ids, metabolites(model)))
+    for idx in proton_idxs
+        JuMP.fix(dgfs[idx], 0.0)
+        JuMP.fix(logcs[idx], 0.0)
     end
 
-    @objective(opt_model, Max, minDF)
+    # set water concentrations to zero (aqueous condition assumptions)
+    water_idxs = Int.(indexin(water_ids, metabolites(model)))
+    for idx in water_idxs
+        JuMP.fix(logcs[idx], 0.0)
+    end
+
+    # set bounds relative to supplied data, fallback to 1% of value
+    dgf_errors = [
+        get(
+            metabolite_gibbs_free_energy_errors,
+            mid,
+            metabolite_gibbs_free_energy_error_fallback_fraction *
+            metabolite_gibbs_free_energies[mid],
+        ) for mid in metabolites(model)
+    ]
```

Comment:
no need to materialize the vector here, just compute the single `dfg_error` before `Jump.set*`